### PR TITLE
Improved tabcompletion and CLI cleanup

### PR DESCRIPTION
**Pull request by stdweird**
Requires https://github.com/quattor/CCM/pull/130

### CODE REVIEW

**Comment by stdweird** [comment]
warning: this runs `ncm-ncd` in tainted mode (https://github.com/quattor/ncm-ncd/pull/81/commits/9257a7e184f1b34f24baaf39d7ff3659ed676200); this might require more testing. i can revert that commit if needed

**Comment by stdweird** [comment]
the `-T` is replaced with with previous `-t`. it was giving some errors on some components.

**Comment by ned21** [comment]
I think this needs a rebase since some prior PRs have been merged. Also please squash 37afd25 into 	4f260b0 so 4f260b0 is a standalone change. I have no experience with tab completion code so I will let someone else review that commit.

**Comment by stdweird** [comment]
@ned21 done.
as soon as #74 is merged, i'll open a new PR on top of this one to handle a bunch of smaller issues

File: src/main/perl/ComponentProxy.pm [hunk 1]
```diff
@@ -494,7 +494,10 @@ sub _execute
     # TODO: return value $result is unused
     local $@;
     my $result;
-    eval "\$result=\$component->$method(\$self->{'CONFIG'});";
+    eval {
+        $component->set_active_config($self->{CONFIG});
```
Comment (piojo) [inline src/main/perl/ComponentProxy.pm]: As discussed in #74 I don't quite like this part.

File: src/main/perl/ComponentProxy.pm [hunk 2]
```diff
@@ -494,7 +494,10 @@ sub _execute
...
+        $result = $component->$method($self->{CONFIG});
```
Comment (piojo) [inline src/main/perl/ComponentProxy.pm]: But this is a nice improvement. :+1:

File: src/main/scripts/ncm-ncd [hunk 3]
```diff
@@ -3,390 +3,35 @@
 # ${developer-info}
 # ${author-info}
 # ${build-info}
-#
-use lib '/usr/lib/perl';
```
Comment (piojo) [inline src/main/scripts/ncm-ncd]: Where are you moving the libraries to?  RHEL5 and 6 had different installation paths for Perl modules.

Comment (piojo) [inline src/main/scripts/ncm-ncd]: +1 for the simplification :smile:

File: src/main/scripts/ncm-ncd.pod [hunk 4]
```diff
@@ -51,32 +51,33 @@ name - active? - installed?
 
 =item --configure
 
-run the 'configure' method for <component1,2..> (default option). For running configure on all components, use --configure --all.
+Run the 'configure' method for <component1 component2 ...> (default option).
```
Comment (piojo) [inline src/main/scripts/ncm-ncd.pod]: `C<configure>` would be nicer POD.  Not important, though.

File: src/test/perl/NCM/Component/foo.pm [hunk 5]
```diff
@@ -12,4 +12,22 @@ our $EC = LC::Exception::Context->new()->will_store_all();
...
+
+    # for unittests
+    # 2 warns, 3 errors
+    $self->warn("foo Configure w1");
+    $self->warn("foo Configure w2");
+    $self->error("foo Configure e1");
+    $self->error("foo Configure e2");
+    $self->error("foo Configure e3");
+
+    $self->{_config} = $config;
```
Comment (piojo) [inline src/test/perl/NCM/Component/foo.pm]: See comments in #74 .

File: src/test/perl/cli.t [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+use Test::More;
+use Test::Quattor::ProfileCache qw(prepare_profile_cache);
+use NCD::CLI;
+use Test::MockModule;
+use LC::Exception;
+
+my $mock_cli = Test::MockModule->new('NCD::CLI');
+my $mock_cpl = Test::MockModule->new('NCD::ComponentProxyList');
+
+$mock_cli->mock('_exit', sub { shift; my $exitcode = shift; die("exit $exitcode");});
```
Comment (piojo) [inline src/test/perl/cli.t]: Looks like you mean

```perl
sub { die "Exit $_[1]"; }
```

File: src/test/perl/component-execute.t [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+# _version_check always ok from now on
+$mock->mock('_version_check', 1);
+
+# multilog
+my $initlog;
+$mock->mock('init_logfile', sub {shift; $initlog = join(" ", @_); return 1;});
+# history
+my $history;
+$mock->mock('set_report_history', sub {shift; $history = 1; return 1;});
```
Comment (piojo) [inline src/test/perl/component-execute.t]: Save yourself the `shift` here.

File: src/test/perl/tabcompletion.t [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+
+
+sub run_bash
+{
+    my ($args, $echo_comp) = @_;
+
+    my $p = CAF::Process->new(['bash', '-c']);
+
+    # Source CCM_TAB_COMP by hand before
+    my @bashargs = ("QUATTOR_CCM_CONF=$cfg->{cache_path}/ccm.conf", 'source', $CCM_TAB_COMP, '&&', 'source', $TAB_COMP, '&&' );
```
Comment (piojo) [inline src/test/perl/tabcompletion.t]: Line too long.

File: src/test/perl/tabcompletion.t [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+    return \@opts;
+}
+
+# from CAF::Application
+Readonly::Array my @CAF_OPTIONS => qw(debug help quiet verbose version);
+is_deeply(bash_split_opts('_quattor_ncm_ncd_default_options'),
+          \@CAF_OPTIONS,
+          "_quattor_ncm_ncd_default_options");
+
+Readonly::Array my @OPTIONS => qw(all allowbrokencomps autodeps cache_root cfgfile check-noquattor chroot configure facility force-quattor forcelock history history-instances ignore-errors-from-dependencies ignorelock include list logdir multilog noaction nodeps post-hook post-hook-timeout pre-hook pre-hook-timeout retries skip state timeout unconfigure useprofile);
```
Comment (piojo) [inline src/test/perl/tabcompletion.t]: Line **way** too long.

**Comment by stdweird** [comment]
@ned21 @piojo remarks addressed, i removed #74 . this still requires https://github.com/quattor/CCM/pull/130

**Pull request merged**